File: src/extensions/groups.ts
import type { CommandSpec } from '../commands/index.js';

export interface CommandGroup {
  name: string;
  commands: readonly CommandSpec[];
}

export type CommandSource = CommandSpec | CommandGroup;

export function defineCommandGroup(name: string, commands: Iterable<CommandSpec>): CommandGroup {
  const trimmedName = name.trim();
  if (trimmedName.length === 0) {
    throw new Error('command group name is required');
  }

  const collectedCommands = Object.freeze([...commands]);
  return Object.freeze({
    name: trimmedName,
    commands: collectedCommands,
  });
}

export function collectCommands(sources: Iterable<CommandSource>): CommandSpec[] {
  const collected: CommandSpec[] = [];

  for (const source of sources) {
    if ('commands' in source) {
      collected.push(...source.commands);
      continue;
    }
    collected.push(source);
  }

  return collected;
}

File: python/src/onetool/extensions/groups.py
from __future__ import annotations

from collections.abc import Iterable
from dataclasses import dataclass
from typing import TypeAlias

from ..commands.core import CommandSpec


@dataclass(frozen=True, slots=True)
class CommandGroup:
    name: str
    commands: tuple[CommandSpec, ...]


CommandSource: TypeAlias = CommandSpec | CommandGroup


def define_command_group(name: str, commands: Iterable[CommandSpec]) -> CommandGroup:
    trimmed_name = name.strip()
    if trimmed_name == "":
        raise ValueError("command group name is required")
    return CommandGroup(name=trimmed_name, commands=tuple(commands))


def collect_commands(sources: Iterable[CommandSource]) -> list[CommandSpec]:
    collected: list[CommandSpec] = []
    for source in sources:
        if isinstance(source, CommandGroup):
            collected.extend(source.commands)
            continue
        collected.append(source)
    return collected
